client/src/main/java/io/pravega/client/stream/impl/ReaderGroupStateManager.java

```diff
@@ -463,7 +470,23 @@ void checkpoint(String checkpointName, PositionInternal lastPosition) throws Rea
     boolean isCheckpointSilent(String atCheckpoint) {
         return sync.getState().getCheckpointState().isCheckpointSilent(atCheckpoint);
     }
-    
+
+    @SneakyThrows(CheckpointFailedException.class)
+    public void updateTruncationStreamCutIfNeeded() {
+        ReaderGroupState state = sync.getState();
+        // here add check if RG is Subscriber and autoPublishTruncationStreamCut = true
+        if (state.getConfig().getRetentionConfig().equals(ReaderGroupConfig.StreamDataRetention.CONSUMPTION_BASED_AT_LAST_CHECKPOINT)
```

Comment:
Why do we check just `CONSUMPTION_BASED_AT_LAST_CHECKPOINT`
Is it invalid to set `CONSUMPTION_BASED_USER_STREAMCUT` and call this method? If so why? There is no inconsistency.
Additionally if the user is making a mistake if this is set to "no_impact" should it silently do nothing or throw?

Reply:
> Is it invalid to set CONSUMPTION_BASED_USER_STREAMCUT and call this method? If so why? 
User needs to set retentionPolicy = 'CONSUMPTION_BASED_USER_STREAMCUT' when user will be providing the truncation StreamCut and does not intend to use the checkpoint StreamCut as truncation point. This has been explained in the enum javadoc

Reply:
> Additionally if the user is making a mistake if this is set to "no_impact" should it silently do nothing or throw?

If the user choses NO_IMPACT, when the stream is configured for Consumption Based Retention, the Consumption Based Retention promise is broken and this is made explicit in the enum javadoc

Reply:
>  Consumption Based Retention promise is broken 

Then the API is broken. It needs to either throw an exception because the case is invalid and document that it does this explicitly, or better yet be written in such a way that invalid states cannot be expressed.

Reply:
> We want to publish the stream-cut corresponding to the last checkpoint only when the user has chosen the later option

Why do we want this? It seems like an application could both ask that the retention cut be updated at checkpoints and sometimes provide its own value. It's certainly possible to write such an application. If the user does that now, they will simply get confusing undocumented behavior. A choice needs to be made as to what it does, and that behavior needs to be defined. If it is "invalid" then it should throw, and the exception should stated in the Javadoc. If it is valid then what happens must be specified.

Reply:
> It seems like an application could both ask that the retention cut be updated at checkpoints and sometimes provide its own value.

This can be disallowed, by the Client API that accepts truncation Stream-Cuts for CBR, by checking  ReaderGroupConfig.retentionConfig == CONSUMPTION_BASED_USER_STREAM_CUT. 

I don;t understand how  the check "retentionConfig == CONSUMPTION_BASED_AT_LAST_CHECKPOINT" breaks anything. It is an absolutely necessary check, because we want to publish that stream-cut only when the user has chosen this option CONSUMPTION_BASED_AT_LAST_CHECKPOINT.

Reply:
> If the user does that now, they will simply get confusing undocumented behavior. 

There is nothing undocumented or confusing here.
The system behaves as per the retentionConfig option chosen by the user.
Please read through the PDP if you have questions/concerns:
https://github.com/pravega/pravega/wiki/PDP-47:-Pravega-Streams:-Consumption-Based-Retention

Reply:
> This can be disallowed, 

The code here is not disallowing this. It should either work and pass the data to the controller or actually enforce that it is disallowed by throwing an exception.

> The system behaves as per the retentionConfig option chosen by the user.

It is not. The docs do not explain what happens if `CONSUMPTION_BASED_USER_STREAMCUT` is specified and this method is invoked. If that is invalid then this must throw. If it is valid it must work. Silently ignoring the call violates either assumption.

Reply:
The updateRetentionStreamCut() call does throw an UnsupportedOperationException when the RG Retention config is not set to "CONSUMPTION_BASED_USER_STREAMCUT".